design-proposals/klip-7-connect-integration.md

```diff
@@ -0,0 +1,205 @@
+this document we will assume that the user table is stored in a JDBC compatible database, as opposed
+to a Kafka changelog topic. After implementing this KLIP, it should be possible for a KSQL user to
+perform a stream-table join on these two without leaving KSQL's console.
+
+## Scope
+
+This KLIP will:
+- define a SQL-like language for basic connector management
+- provide a guiding principles and high level architecture for a connect-ksql integration
+
+This KLIP will _not_:
+- propose improvements to Kafka Connect such as easing the configuration burden, but will discuss
+  how to leverage future Connect improvements
+
+## Value/Return
+
+This proposal targets _usability_ improvements for KSQL and simplifies bootstrapping KSQL for users
+that may already have some, but not all, of their data in Kafka. As with the motivating example,
+KSQL users will be able to easily model external systems within KSQL. The target user has already
+played around with KSQL and is working on prototyping their first real application with existing
+data.
+
+## Design
+
+### Architecture
+
+There are three ways to integrate KSQL with Connect:
+1. Run them separately and supply syntax to communicate between them
+1. Embed Connect into KSQL, managing individual connect workers as part of the KSQL ecosystem
+1. Package Connect and KSQL into a single JVM, but just spin them up at the top level
+
+There is some consensus that at scale KSQL and Connect should run as separate components. This
+means that the API that we design will not be able to leverage any "embedded-ness" of Connect,
+nullifying most of the benefits of the second approach. 
+
+There are still, however, benefits to packaging connect and KSQL in the same JVM for a seamless
+experience for new users. Namely, new users will only need to download a single package, spin up
+a single process and maintain only a single application. There are some (solvable) concerns with
+this approach, but for the reason outlined above, the syntax will be developed with the assumption
+that Connect is running as a separate component.
+
+### Syntax
+
+Analogous to DDL and DML, the Connect syntax will separate actions that have side effects with
+actions that simply declare metadata on top of existing data. This solves two issues:
+
+- A single connector can have many topics, which makes modeling them as a single statement difficult
+- A connector can be modeled in different ways (e.g. stream/table) and connector lifecycle should
+  be independent with how KSQL interprets it
+
+In the future, there can exist shortcuts (similar to `C*AS`) to both create connectors and declare 
+corresponding streams/tables.
+
+There are a few more minor points of consideration:
+- All "DDL" commands that cause changes in connect will not be submitted to the command topic;
+  instead they will be issued on the REST server (this follows the pattern that all external-state
```

Comment:
An alternative to having a KSQL cluster know about a Connect cluster is to have the same logic in the CLI.  This has the benefit that we keep the KSQL server and Connect decoupled.

Aside from keeping the KSQL server code cleaner this also means a different users / teams can use different Connect clusters with the same KSQL cluster, i.e. the mapping changes from being:

> A KSQL cluster is configured to work with a connect cluster

To

> A KSQL CLI is configured to work with a KSQL cluster and a Connect cluster.

And we can have CLI commands to switch Connect cluster, as we do for KSQL cluster.

In effect, I'm proposing the integration point for Connect/KSQL should be in the CLI / UI, not in the KSQL server.

Reply:
I'm a little nervous about the portability of that approach, though I don't feel that strongly about it. What if I wanted to start my connectors as part of my sandbox deployment (e.g. headless mode)?

Reply:
We can also have a rest API that knows how to deal with both, and forward statements to either the Connect RESTful API or the KSQL RESTful API appropriately.  What I think we should avoid is polluting the KSQL RESTFul API / Engine with Connect specific code.  We need to maintain good separation of concern / architectural boundaries.

Reply:
I agree with this design goal and will keep it in mind during implementation. I have a preference toward not having the CLI communicate directly with another service (especially when things like endpoint discovery become problematic), but we should definitely keep these distinct and modular.

Reply:
One good reason to integrate the Connect code in the CLI is around security. The user/password of the databases supplied won't go to the KSQL server which can be seen by the KSQL service principal (if we don't trust on the KSQL service principal of course), instead, they will go directly to Connect. However, a drawback is what if Connect is open for connections from the KSQL server only and not from any other client? That would not allow our CLI to communicate with Connect. It would be good to learn about how users would use Connect and KSQL and how they will configure it, right?